crash_rate_aggregates/fill_database.py

```diff
@@ -0,0 +1,192 @@
+                (ping["payload/keyedHistograms/SUBPROCESS_CRASHES_WITH_DUMP/plugin_parent"] or 0), # plugin crashes
+                (ping["payload/keyedHistograms/SUBPROCESS_CRASHES_WITH_DUMP/gmplugin_parent"] or 0), # GMplugin crashes
+            ])
+        )
+    crash_values = ping_properties.filter(is_valid).map(get_crash_pair).reduceByKey(lambda a, b: a + b)
+
+    def dimension_mapping(pair): # responsible for converting aggregate crash pairs into individual dimension fields
+        dimension_key = pair[0]
+        (submission_date, activity_date), dimension_values = dimension_key[:2], dimension_key[2:]
+        usage_hours, main_crashes, content_crashes, plugin_crashes, gmplugin_crashes = pair[1]
+        return (
+            submission_date, activity_date,
+            {
+                key: str(dimension_value) if dimension_value is not None else ""
+                for key, dimension_value in zip(dimension_names, dimension_values)
+            },
+            {
+                "usage_hours": float(usage_hours),
+                "main_crashes": float(main_crashes),
+                "content_crashes": float(content_crashes),
+                "plugin_crashes": float(plugin_crashes),
+                "gmplugin_crashes": float(gmplugin_crashes),
+            },
+        )
+    return crash_values.map(dimension_mapping)
+
+def retrieve_crash_data(sc, submission_date_range, comparable_dimensions, fraction = 0.1):
+    # get the raw data
+    normal_pings = get_pings(
+        sc, doc_type="main",
+        submission_date=submission_date_range,
+        fraction=fraction
+    )
+    crash_pings = get_pings(
+        sc, doc_type="crash",
+        submission_date=submission_date_range,
+        fraction=fraction
+    )
+
+    return normal_pings.union(crash_pings)
+
+def run_job(spark_context, sql_context, submission_date_range):
+    """Fill the specified database with crash aggregates for the specified submission date range, creating the schema if needed."""
+    start_date = datetime.strptime(submission_date_range[0], "%Y%m%d").date()
+    end_date = datetime.strptime(submission_date_range[1], "%Y%m%d").date()
+
+    schema = types.StructType([
+        types.StructField("activity_date",   types.DateType(),                                            nullable=False),
+        types.StructField("dimensions",      types.MapType(types.StringType(), types.StringType(), True), nullable=False),
+        types.StructField("stats",           types.MapType(types.StringType(), types.DoubleType(), True), nullable=False),
+    ])
+
+    current_date = start_date
+    while current_date <= end_date:
+        # useful statements for testing the program
+        #import sys, os; sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "test")); import dataset; pings = sc.parallelize(list(dataset.generate_pings())) # use test pings; very good for debugging queries
```

Comment:
This one will upload the results to S3, which is still pretty useful if we're changing the upload functionality. It's not really practical to do that part in tests since only one person could do so at a time, or we'd need to upload to a different place for each user.